src/version.rs

```diff
@@ -404,6 +515,143 @@ pub fn downloads(req: &mut Request) -> CargoResult<Response> {
+            beta: self.beta.map(::encode_time),
+            nightly: self.nightly.map(::encode_time),
+        }
+    }
+}
+
+#[derive(Debug, Default, RustcEncodable, RustcDecodable)]
+pub struct EncodableMaxBuildInfo {
+    pub stable: Option<String>,
+    pub beta: Option<String>,
+    pub nightly: Option<String>,
+}
+
+impl BuildInfo {
+    pub fn max<I>(build_infos: I) -> CargoResult<MaxBuildInfo>
+        where I: IntoIterator<Item = BuildInfo>
+    {
+        let mut stable = MaxOrNone::new();
+        let mut beta = MaxOrNone::new();
+        let mut nightly = MaxOrNone::new();
+
+        for row in build_infos {
+            let BuildInfo { rust_version, .. } = row;
+            let rust_version: ChannelVersion = rust_version.parse()?;
+
+            match rust_version {
+                ChannelVersion::Stable(semver) => stable.push(semver),
+                ChannelVersion::Beta(date) => beta.push(date),
+                ChannelVersion::Nightly(date) => nightly.push(date),
+            }
+        }
+
+        Ok(MaxBuildInfo {
+            stable: stable.0,
+            beta: beta.0,
+            nightly: nightly.0,
+        })
+    }
+}
+
+/// Handles the `GET /crates/:crate_id/:version/build_info` route.
+pub fn build_info(req: &mut Request) -> CargoResult<Response> {
+    let (version, _) = try!(version_and_crate(req));
+
+    let conn = req.db_conn()?;
+
+    let build_infos = BuildInfo::belonging_to(&version)
+        .select(BUILD_INFO_FIELDS)
+        .load(&*conn)?;
+
+    let mut encodable_build_info = EncodableBuildInfo::default();
+    encodable_build_info.id = version.id;
+    let mut stables = BTreeSet::new();
+    let mut betas = BTreeSet::new();
+    let mut nightlies = BTreeSet::new();
+
+    for row in build_infos {
+        let BuildInfo { rust_version, target, passed, .. } = row;
+        let rust_version: ChannelVersion = rust_version.parse()?;
+
+        match rust_version {
+            ChannelVersion::Stable(semver) => {
+                let key = semver.to_string();
+                stables.insert(semver);
+                encodable_build_info.stable.entry(key)
+                                 .or_insert_with(HashMap::new)
+                                 .insert(target, passed);
+            }
+            ChannelVersion::Beta(date) => {
+                let key = ::encode_time(date);
+                betas.insert(date);
+                encodable_build_info.beta.entry(key)
+                                  .or_insert_with(HashMap::new)
+                                  .insert(target, passed);
+            }
+            ChannelVersion::Nightly(date) => {
+                let key = ::encode_time(date);
+                nightlies.insert(date);
+                encodable_build_info.nightly.entry(key)
+                                  .or_insert_with(HashMap::new)
+                                  .insert(target, passed);
+            }
+        }
+    }
+
+    encodable_build_info.ordering.insert(
```

Comment:
We wouldn't need to maintain this `ordering` field, since `OrderMap` preserves insertion order. Even if we still need `ordering` for the front-end, it'd just be `.keys().collect()` to get the corresponding vec.

Though on second reading, that's not even what this code is doing. Since it's in a `BTreeSet`, it's just going to be sorting the strings in ascending order. Is that the desired behavior? It seems like we could just sort them when we're ready to display them in that case.